GH-1387 Improved custom service executor extension system

Aklakan opened the pull request:
GitHub issue resolved #1387.
This fixes #1399.

Pull request Description: This PR adds the following changes to the custom service executor plugin system:

Improvements:

* **chaining** There is now `ChainingServiceExecutor` which allows for transparently forwarding a request to the next `ChainingServiceExecutor` instance in the registry. This way a custom service executor can modify the request and have it processed by the remainder of the chain.
* **bulk processing** The plugin system introduced with jena 4.5.0 only supports lookups with individual bindings. This PR adds support for plugins that want to process bindings in bulk. The main difference in the APIs is whether the `createExecution` method accepts a single `Binding` or a `QueryIterator`. There is now a `ServiceExecutorRegistryBulk` which by default has an registration that delegates to the non-bulk one.

Breaking changes:

* `ServiceExecutorRegistry.getFactories()` now returns a `List<ChainingServiceExecutor>` because this is its internal storage; previously it was `List<ServiceExecutorFactory>`. I am afraid re-establishing compatibility in this regard would require a completely new registry with a new context attribute.

Deprecations:
* Deprecated `ServiceExecutorFactory` in favor of `ServiceExecutor` (ServiceExecutorFactory extends from the latter). Use of the old code delegates to the new one.
    * In the new code the method that creates a `ServiceExecution` is now called `createExecution` (rather than `createExecutor`)
* ~Deprecated `ServiceExecutorRegistry.add` in favor of `prepend`. Both methods add executors to the beginning of the registry's list so they are considered before the default behavior.~ QueryEngineFactory also uses `add` to prepend; so `add` is consistent with existing naming.
* Legacy add method for `ServiceExecutorFactory` exists and wraps it as a `ChainingServiceExecutor`.
* Legacy remove method for `ServiceExecutorFactory` finds previously wrapped elements..
* Deprecated `ServiceExecution` because it doesn't add anything over `QueryIterator`.

----

 - [x] Javadoc exists and is up-to-data
 - [x] Tests are included.
 - [ ] Documentation change and updates are provided for the [Apache Jena website](https://github.com/apache/jena-site/)
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx or JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterRepeatApply.java:
```diff
@@ -18,99 +18,26 @@
-            return null;
-        }
-
+    protected QueryIterator nextStage(QueryIterator input) {
```
Please explain what is going on here. 

If a change is needed, then having a superclass is not necessary. Just two `nextStage` methods with different signatures.

Aklakan replied:
QueryIterApply**Bulk**'s `nextStage(QueryIterator)` method flat-maps the input `QueryIterator` to an output `QueryIterator` until the input iterator is consumed - i.e. implementations of nextStage can "per stage" consume an arbitrary amount > 0 from the input iterator.

QueryIterApply implements `QueryIterApplyBulk.nextStage(QueryIterator)` such that it delegates to its own usual `QueryIterApply.nextStage(Binding)` method.

Note that `QueryIterApplyBulk` is also the base class for `QueryIterServiceBulk`. The latter only operates on the `QueryIterator` and therefore does not need a `nextStage(Binding)` method. I find this separation cleaner than having both `nextStage` methods on the same class together with an override that causes the binding-level one to no longer be called.

The only drawback I see with this separation is that reflection code that relied on `QueryIterApply.class.getDeclaredMethods()` would break.

afs replied:
> I find this separation cleaner

My concern is that this is making the bulk case "normal" when in fact it is special to your service work.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/ARQConstants.java:
```diff
@@ -292,6 +292,10 @@
     public static final Symbol registryServiceExecutors =
         SystemARQ.allocSymbol("registryServiceExecutors") ;
 
+    /** The service executor library registry key */
```
Are two different keys necessary?

Aklakan replied:
Very good question. Right now the bulk and non-bulk registries are two separate classes. I need to think about whether they could be combined into a single class.

Aklakan commented:
Just to clarify, my next step on this PR is to update and write javadoc; especially for what some of the classes are for. This should make reviewing easier.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/iterator/QueryIterRepeatApply.java:
```diff
@@ -18,99 +18,26 @@
 
     protected abstract QueryIterator nextStage(Binding binding);
 
+    /** The caller guarantees that input has at least one more item */
```
That is not the contract of `hasNextBinding`.

Aklakan commented:
May latest attempt at simplifying things:

* Reverted QueryIterRepeatApply (it used inside of ServiceExecutorBulkToSingle which acts as the bridge between bulk and non-bulk)
* Combined bulk and non-bulk registries
* Deprecated ServiceExecution in favor using QueryIterator directly
* OpExecutor now directly calls the plugin system
* This way QueryIterService is actually not used anymore - service execution via the registry eventually by default ends up with a QueryIterRepeatApply. The still existing QueryIterService only uses the non-bulk part of the registry.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/QueryIterator.java:
```diff
@@ -18,15 +18,13 @@
 /** Root of query iterators in ARQ. */
 
-public interface QueryIterator extends Closeable, Iterator<Binding>, PrintSerializable
+public interface QueryIterator extends ClosableIterator<Binding>, PrintSerializable
```
This change isn't necessary but if it is made, please use `IteratorCloseable`  See the comments in ClosableIterator which refer to the Model API and `ExtendedIterator` - let's keep `ClosableIterator` for jena-core.

afs commented on jena-arq/src/main/java/org/apache/jena/sparql/engine/main/iterator/QueryIterService.java:
```diff
@@ -32,8 +32,8 @@
 import org.apache.jena.sparql.engine.main.QC ;
 import org.apache.jena.sparql.exec.http.Service;
 import org.apache.jena.sparql.service.ServiceExecution;
```
This is now an unused import and can be removed.

Aklakan replied:
I removed the import and added a deprecation annotation because QueryIterService is now no longer used but existing code may have sub-classed from it.

afs commented:
Looks good - 2 small comments.

Then it's the checklist items : javadoc item and how you want to commits to look when merged to the codebase. e.g any squashing and and any use of "GH-1387: " to pick out key commits.

Aklakan commented:
I noticed that the overloaded `add` methods in ServiceExecutorRegistry would result in ambiguity with lambdas - which broke with (my) code based on the 4.2.0 api.
Before the naming gets frozen I now renamed the methods to:
```
add(ServiceExecutor) // Compatible with 4.2.0
addSingleLink(ChainingServiceExecutor) // New method; adds a link to the chain
addBulkLink(ChainingServiceExecutorBulk) // Also new
```
I updated javadoc and added a couple of ```@deprecated(since = "4.6.0")``` annotations.

afs commented:
Are there any documentation changes for `jena-site`?
